java/src/main/java/ai/rapids/cudf/HostColumnVector.java

```diff
@@ -738,15 +738,49 @@ public ColumnBuilder appendLists(List... inputLists) {
     public ColumnBuilder appendStructValues(List<StructData> inputList) {
       for (StructData structInput : inputList) {
         // one row
-        rows++;
         append(structInput);
       }
       return this;
     }
 
+    private void resizeDataBuffer(int length) {
+      // just for strings we want to throw a real exception if we would overrun the buffer
+      if (data == null) {
+        // since rows can be updated as we go, make data buffer as big as the estimate or the requested length
+        if (!type.isNestedType() && type != DType.STRING) {
+          data = HostMemoryBuffer.allocate(Math.min(Integer.MAX_VALUE, estimatedRows * type.getSizeInBytes()));
+        } else {
+          data = HostMemoryBuffer.allocate(length);
+        }
+        return;
+      }
+      long oldLen = data.getLength();
+      long newLen = Math.max(1, oldLen);
+      while (currentByteIndex + length > newLen) {
+        newLen *= 2;
+      }
+      if (newLen > Integer.MAX_VALUE) {
+        throw new IllegalStateException("A data buffer is not supported over 2GB in size");
```

Comment:
On a separate note even for bytes this is wrong because we are now combining estimatedRows with doubling each time, so we could go over the size limit on a string column when all we needed was an extra byte.  I'd prefer to see us clamp newLen to a pre-calculated maximum length and then check if the desired extra length would go over the maximum length.

Reply:
Yes that is true. I will clip it and make it more precise like offsets and validity.

Reply:
I have documented based on my understanding how we do data buffers sizing now. Let me know if you think there is more here.